compiler/rustc_infer/src/infer/sub.rs

```diff
@@ -141,17 +142,27 @@ impl<'tcx> TypeRelation<'tcx> for Sub<'_, '_, 'tcx> {
                     Ok(infcx.tcx.mk_ty_var(var))
                 };
                 let (a, b) = if self.a_is_expected { (a, b) } else { (b, a) };
-                let (a, b) = match (a.kind(), b.kind()) {
+                let (ga, gb) = match (a.kind(), b.kind()) {
                     (&ty::Opaque(..), _) => (a, generalize(b, true)?),
                     (_, &ty::Opaque(..)) => (generalize(a, false)?, b),
                     _ => unreachable!(),
                 };
                 self.fields.obligations.extend(
                     infcx
-                        .handle_opaque_type(a, b, true, &self.fields.trace.cause, self.param_env())?
+                        .handle_opaque_type(ga, gb, true, &self.fields.trace.cause, self.param_env())
+                        // Don't leak any generalized type variables out of this
+                        // subtyping relation in the case of a type error.
+                        .map_err(|err| {
+                            let (ga, gb) = self.fields.infcx.resolve_vars_if_possible((ga, gb));
```

Comment:
This is probably overkill, and we maybe should just do something like `.map_err(|_| TypeError::Sorts(ExpectedFound { expected: a, found: b }))` instead.

Reply:
yea, I was just checking out all error sources in `handle_opaque_types`, and they all boil down to calling `infcx.at(...).eq(x, y)?` in some manner. The issue is that sometimes it's not `a` and `b` as passed, but another previously registered hidden type. Then you want to report the previous hidden type mismatching the new hidden type.

So I guess I'm saying there is a version of this issue where your change will actually not catch it. Trying to produce an example now

Reply:
I was unable to reproduce the issue in the defining scope (that doesn't mean it's not possible, just that I gave up for now ^^). I found this absolutely useless diagnostic though: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=b80473c86172959763c139535154c3f9

Did some more thinking and I'm wondering now if the `did.is_local()` check should actually be `infcx.opaque_type_origin(did).is_some()` so we only generate those generalized inference vars if we're actually able to handle the opaque type. I was kinda hoping to keep that isolated in `handle_opaque_type`, but that appears to have been a misplaced hope.

One more thing is that we're using `eq` inside `handle_opaque_types` when it should probably be using the relation that is calling `handle_opaque_types`. So maybe we could scrap the entire generalization logic and just always invoke `handle_opaque_types` by passing the relation as an argument and forwarding to that instead of using `eq`.

so many possibilities to try out. Let's merge this PR for now and I'm moving this comment to an issue so we can continue there.